src/lib/io/sol-memmap-storage-impl-zephyr.c

```diff
@@ -0,0 +1,285 @@
+    if (r < 0) {
+        SOL_WRN("Flash erase (before write) failed");
+        goto error;
+    }
+
+    if (mask) {
+        uint64_t value = 0, old_value;
+        uint32_t i, j;
+
+        /* entry->size > 8 implies that no mask should be used */
+        assert(entry->size <= 8);
+
+        for (i = 0, j = 0; i < entry->size; i++, j += 8)
+            value |= (uint64_t)((uint8_t *)blob->mem)[i] << j;
+
+        memcpy(&old_value, back_mem + mod, entry->size);
+
+        value <<= entry->bit_offset;
+        value &= mask;
+        value |= (old_value & ~mask);
+
+        memcpy(back_mem + mod, &value, entry->size);
+
+        r = flash_write_protection_set(flash_dev, false);
+        SOL_INT_CHECK_GOTO(r, < 0, error);
+
+        for (cnt = 0; cnt < back_sz; cnt += MAX_RW) {
+            r = flash_write(flash_dev, offset - mod + cnt, back_mem + cnt,
+                MAX_RW);
+            if (r < 0) {
+                SOL_WRN("Flash write failed");
+                goto error;
+            }
+            r = flash_write_protection_set(flash_dev, false);
+            SOL_INT_CHECK_GOTO(r, < 0, error);
+        }
+    } else {
+        size_t size;
+
+        if (blob->size > entry->size) {
+            SOL_WRN("Trying to store entry data of size %zu bytes to a entry "
+                "with %zu bytes of reserved space", blob->size, entry->size);
+        }
+        size = sol_util_min(entry->size, blob->size);
+
+        memcpy(back_mem + mod, blob->mem, size);
+
+        r = flash_write_protection_set(flash_dev, false);
+        SOL_INT_CHECK_GOTO(r, < 0, error);
+
+        for (cnt = 0; cnt < back_sz; cnt += MAX_RW) {
+            r = flash_write(flash_dev, offset - mod + cnt, back_mem + cnt,
+                MAX_RW);
+            if (r < 0) {
+                SOL_WRN("Flash write failed");
+                goto error;
+            }
+            r = flash_write_protection_set(flash_dev, false);
+            SOL_INT_CHECK_GOTO(r, < 0, error);
+        }
+    }
+
+    if (cb)
+        cb((void *)data, name, blob, r);
+
+    sol_blob_unref(blob);
+
+    return r;
+
+error:
+    SOL_DBG("Error writing to memmap: %s", sol_util_strerrora(-r));
+
+    if (cb)
+        cb((void *)data, name, blob, r);
+
+    sol_blob_unref(blob);
+
+    return r;
+}
+
+bool
+sol_memmap_impl_perform_pending_writes(void *data)
+{
+    struct map_internal *map_internal = data;
+    struct pending_write_data *pending;
+    struct sol_vector tmp_vector;
+    int i;
+
+    map_internal->timeout = NULL;
+
+    tmp_vector = map_internal->pending_writes;
+    sol_vector_init(&map_internal->pending_writes,
+        sizeof(struct pending_write_data));
+
+    SOL_VECTOR_FOREACH_IDX (&tmp_vector, pending, i) {
+        sol_memmap_impl_write_raw((struct map_internal *)map_internal,
+            pending->name, pending->entry, pending->mask, pending->blob,
+            pending->cb, pending->data);
+        free(pending->name);
```

Comment:
I have to admit I didn't look in detail, but I wonder if we should look that in the memmap table and use that name instead, then just point to an existing string instead of strdup one